language/vm/vm-runtime/vm-runtime-types/src/values/values_impl.rs

```diff
@@ -0,0 +1,2223 @@
+
+    pub const INDEX_OUT_OF_BOUNDS: u64 = NFE_VECTOR_ERROR_BASE + 1;
+    pub const POP_EMPTY_VEC: u64 = NFE_VECTOR_ERROR_BASE + 2;
+    pub const DESTROY_NON_EMPTY_VEC: u64 = NFE_VECTOR_ERROR_BASE + 3;
+
+    macro_rules! ensure_len {
+        ($v: expr, $expected_len: expr, $type: expr, $fn: expr) => {{
+            let actual_len = $v.len();
+            let expected_len = $expected_len;
+            if actual_len != expected_len {
+                let msg = format!(
+                    "wrong number of {} for {} expected {} found {}",
+                    ($type),
+                    ($fn),
+                    expected_len,
+                    actual_len,
+                );
+                return Err(VMStatus::new(StatusCode::UNREACHABLE).with_message(msg));
+            }
+        }};
+    }
+
+    macro_rules! pop_arg_front {
+        ($arguments:ident, $t:ty) => {
+            $arguments.pop_front().unwrap().value_as::<$t>()?
+        };
+    }
+
+    macro_rules! err_vector_elem_ty_mismatch {
+        () => {{
+            return Err(VMStatus::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                .with_message("vector elem type mismatch".to_string()));
+        }};
+    }
+
+    pub fn native_empty(
+        ty_args: Vec<TypeTag>,
+        args: VecDeque<Value>,
+        cost_table: &CostTable,
+    ) -> VMResult<NativeResult> {
+        ensure_len!(ty_args, 1, "type arguments", "empty");
+        ensure_len!(args, 0, "arguments", "empty");
+
+        let cost = native_gas(cost_table, NativeCostIndex::EMPTY, 1);
+        let container = match &ty_args[0] {
+            TypeTag::U8 => Container::U8(vec![]),
+            TypeTag::U64 => Container::U64(vec![]),
+            TypeTag::U128 => Container::U128(vec![]),
+            TypeTag::Bool => Container::Bool(vec![]),
+
+            _ => Container::General(vec![]),
+        };
+
+        Ok(NativeResult::ok(
+            cost,
+            vec![Value(ValueImpl::new_container(container))],
+        ))
+    }
+
+    pub fn native_length(
+        ty_args: Vec<TypeTag>,
+        mut args: VecDeque<Value>,
+        cost_table: &CostTable,
+    ) -> VMResult<NativeResult> {
+        ensure_len!(ty_args, 1, "type arguments", "length");
+        ensure_len!(args, 1, "arguments", "length");
+
+        let cost = native_gas(cost_table, NativeCostIndex::LENGTH, 1);
+        let r = pop_arg_front!(args, ContainerRef);
+        let v = r.try_borrow()?;
+
+        let len = match (&ty_args[0], &*v) {
+            (TypeTag::U8, Container::U8(v)) => v.len(),
+            (TypeTag::U64, Container::U64(v)) => v.len(),
+            (TypeTag::U128, Container::U128(v)) => v.len(),
+            (TypeTag::Bool, Container::Bool(v)) => v.len(),
+
+            (TypeTag::Struct(_), Container::General(v))
+            | (TypeTag::ByteArray, Container::General(v))
+            | (TypeTag::Address, Container::General(v)) => v.len(),
+
+            _ => err_vector_elem_ty_mismatch!(),
```

Comment:
is your idea that when we do the work for Vector we actually define a specific enum variant for vectors?
I mean this code should not fail if it is a vector, or when does it fail?